src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -0,0 +1,161 @@
+            {
+                lock (_gate)
+                {
+                    if (_documentIdToLastResultId.TryGetValue((workspace, document.Id), out var lastReportedResultId) &&
+                        lastReportedResultId == documentDiagnosticsParams.PreviousResultId)
+                    {
+                        // Nothing changed between the last request and this one.  Report a null response to the client
+                        // to know they don't need to do anything.
+                        return null;
+                    }
+                }
+            }
+
+            // Being asked about this document for the first time.  Or being asked again and we have different diagnostics.
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            using var _ = ArrayBuilder<LspDiagnostic>.GetInstance(out var result);
+            foreach (var diagnostic in _diagnosticService.GetDiagnostics(document, includeSuppressedDiagnostics: false, cancellationToken))
+                result.Add(Convert(text, diagnostic));
+
+            lock (_gate)
+            {
+                // Keep track of the diagnostics we reported here so that we can short-circuit producing diagnostics for
+                // the same diagnostic set in the future.
+                var resultId = _nextResultId++.ToString();
+                _documentIdToLastResultId[(workspace, document.Id)] = resultId;
+                return new[] { new DiagnosticReport { ResultId = resultId, Diagnostics = result.ToArray() } };
+            }
+        }
+
+        private static VSDiagnostic Convert(SourceText text, DiagnosticData diagnosticData)
+        {
+            Contract.ThrowIfNull(diagnosticData.Message, $"Got a document diagnostic that did not have a {nameof(diagnosticData.Message)}");
+            Contract.ThrowIfNull(diagnosticData.DataLocation, $"Got a document diagnostic that did not have a {nameof(diagnosticData.DataLocation)}");
+
+            return new VSDiagnostic
+            {
+                Code = diagnosticData.Id,
+                Message = diagnosticData.Message,
+                Severity = ProtocolConversions.DiagnosticSeverityToLspDiagnositcSeverity(diagnosticData.Severity),
+                Range = ProtocolConversions.LinePositionToRange(DiagnosticData.GetLinePositionSpan(diagnosticData.DataLocation, text, useMapped: true)),
+                Tags = ConvertTags(diagnosticData),
+                DiagnosticType = diagnosticData.Category,
+            };
+        }
+
+        private static DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData)
+        {
+            using var _ = ArrayBuilder<DiagnosticTag>.GetInstance(out var result);
+
+            if (diagnosticData.Severity == DiagnosticSeverity.Hidden)
+            {
+                result.Add(VSDiagnosticTags.HiddenInEditor);
+                result.Add(VSDiagnosticTags.HiddenInErrorList);
```

Comment:
Does this also need ```SuppressEditorToolTip``` to ensure hovering over the location doesn't show the hidden diagnostic message?

Reply:
good point :)